### PR TITLE
BWVONE-48664: fix request_parameter advanced setting, and import tool to use proper version

### PR DESCRIPTION
**Pull request by amondal-akam**
BWVONE-48664: fix request_parameter advanced setting
BWVONE-48214: fix import tool to use proper version

### CODE REVIEW

File: tools/utils.go [hunk 1]
```diff
@@ -15,6 +15,8 @@ import (
 	"github.com/akamai/AkamaiOPEN-edgegrid-golang/v6/pkg/edgegrid"
 )
 
+var version = "dev" // overridden at build time via -ldflags
```
[the remark this answers is not in the document]
Comment (amondal-akam) [reply tools/utils.go:18]: Let this break. 
Ideally version would be picked up by Makefile, if its not picked up then it means its not working.

File: pkg/eaaprovider/app_read_helpers.go [hunk 2]
```diff
@@ -347,13 +347,9 @@ func mapAdvancedSettingsFromResponse(d *schema.ResourceData, appResp *client.App
...
-	// request_parameters: map -> JSON string; only set when non-empty (nil means absent).
-	if len(appResp.AdvancedSettings.RequestParameters) > 0 {
-		rpJSON, err := json.Marshal(appResp.AdvancedSettings.RequestParameters)
-		if err != nil {
-			return diag.Errorf("failed to marshal request_parameters to JSON: %v", err)
-		}
-		full["request_parameters"] = string(rpJSON)
+	// request_parameters: *string from API; set as-is when non-empty (nil means absent).
+	if appResp.AdvancedSettings.RequestParameters != nil && *appResp.AdvancedSettings.RequestParameters != "" {
+		full["request_parameters"] = *appResp.AdvancedSettings.RequestParameters
```
[the remark this answers is not in the document]
Comment (amondal-akam) [reply pkg/eaaprovider/app_read_helpers.go:354]: server-side won't be changing it, request_parameters is legacy advanced setting.

File: pkg/client/app_advanced_settings_test.go [hunk 3]
```diff
@@ -175,23 +175,10 @@ func TestAdvancedSettingsFromBlockDecodesRequestParameters(t *testing.T) {
...
-	_, err := advancedSettingsFromBlock(block)
-	if err == nil {
-		t.Fatal("expected error for invalid request_parameters JSON, got nil")
-	}
-	if !strings.Contains(err.Error(), "invalid request_parameters JSON") {
-		t.Fatalf("error = %q, want message containing %q", err.Error(), "invalid request_parameters JSON")
+	if *got.RequestParameters != `{"key":"value"}` {
+		t.Fatalf("RequestParameters = %q, want %q", *got.RequestParameters, `{"key":"value"}`)
 	}
 }
```
[the remark this answers is not in the document]
Comment (amondal-akam) [reply pkg/client/app_advanced_settings_test.go:183]: This is fine for now.

Unit tests need to be improved and revamped in a different PR.

File: pkg/eaaprovider/app_read_helpers.go [hunk 4]
```diff
@@ -421,12 +421,22 @@ func mapAgentsAndAuthFromResponse(d *schema.ResourceData, appResp *client.Applic
...
 				return diag.FromErr(err)
 			}
 		}
+	} else {
+		err = d.Set("app_authentication", []map[string]interface{}{{
+			"app_idp":         "",
+			"app_directories": []map[string]interface{}{},
+		}})
+		if err != nil {
+			return diag.FromErr(err)
```
[the remark this answers is not in the document]
Comment (amondal-akam) [reply pkg/eaaprovider/app_read_helpers.go:438]: when AuthEnabled is false, the API automatically makes  app_idp, and app_directories nil/[]

**Review by ak-amitpathak — COMMENTED** [review]
## Code Review

**Tickets:** BWVONE-48664, BWVONE-48214

### Overview

Two fixes bundled together:
1. **BWVONE-48664**: `request_parameters` was wrongly typed as `map[string]interface{}` — the EAA API actually stores it as a raw JSON string. Fix changes the struct field to `*string` and removes the intermediate marshal/unmarshal layer.
2. **BWVONE-48214**: The import tool was hard-coding version `"1.0.0"` in generated `.tf` files. Fix injects the binary version at build time via `-ldflags`.

Both fixes are directionally correct. A few items need attention.

---

### Correctness

**`request_parameters` type change (`application.go:1840`, `application.go:2001`)**
- Changing `map[string]interface{}` → `*string` is the right fix if the API stores this field as a JSON string natively. The old code was encoding a map to JSON on write and decoding it from JSON on read — a round-trip that could silently corrupt field ordering or numeric types.
- **Concern**: `advancedSettingsFromBlock` uses reflection to populate the `AdvancedSettings` struct from the Terraform state map (`flat`). The deleted special-case code handled `request_parameters` explicitly before reflection ran. Now reflection must assign a `string` value from the flat map to a `*string` struct field. This works only if `applyAdvancedSettingsWithReflection` already handles `string → *string` coercion. **This path is not covered by the updated tests** — the test only checks the read path. Worth verifying manually or adding a write-path test.

**`appauthmembership.go` — `app_directories` always set**
- Moving `appAuth["app_directories"] = directoriesData` outside the `len > 0` guard means an empty slice is now always sent to the API. This is better for state consistency, but **verify the EAA API tolerates `"app_directories": []` vs omission** — some APIs treat an absent key differently from an empty array on PATCH/PUT.
- Setting `app_idp: ""` and `app_directories: []` when `appIDPMembership == nil` is a good defensive addition to prevent state drift.

**`app_read_helpers.go:421` — `authErr == nil` → `if authErr != nil`**
- The functional logic for the success path (`d.Set(...)`) is unchanged — it still runs when `authErr == nil`. The improvement is adding error logging for the failure case and the new `else` block that sets empty auth defaults when auth is disabled. Not a logic bug fix, but a meaningful quality improvement.

**`app_read_helpers.go:357` — write `""` only when key already in state**
- The `existingKeys["request_parameters"]` guard is correct to avoid introducing phantom state on import. Good pattern.

---

### Code Quality

**`tools/utils.go` — HCL indentation fix**
- The old `writeProviderBlock` used tab indentation inside a backtick string, which produced malformed HCL (Terraform uses 2-space). The fix is correct and overdue.

**`tools/utils.go:16` — `var version = "dev"`**
- Clean pattern for build-time injection. The `"dev"` fallback is appropriate. One minor point: the docs bump to `2.0.0` and the tool writes `version` at build time — confirm `VERSION_STR` in the Makefile actually evaluates to `2.0.0` (or whatever the current release is) so generated configs don't mismatch.

**Removed test `TestAdvancedSettingsFromBlockErrorOnInvalidRequestParameters`**
- Correct removal — since `request_parameters` is no longer parsed, invalid JSON is not an error at this layer.

---

### Risks

| Risk | Severity | Notes |
|---|---|---|
| Reflection `string → *string` for `request_parameters` not tested on write path | Medium | Could silently leave field unset on CREATE/UPDATE |
| `app_directories: []` always sent — API may reject empty array | Low–Medium | Check EAA API contract for this field |
| `VERSION_STR` in Makefile vs `2.0.0` docs version may drift | Low | Cosmetic but confusing for users |

---

### Suggestions

1. **Add a write-path test** that calls `advancedSettingsFromBlock` with `request_parameters: '{"key":"value"}'` and asserts the resulting `AdvancedSettings.RequestParameters` is a non-nil `*string` with the expected value — distinct from the existing read test.
2. **Add a test for `CreateAppAuthenticationStruct`** covering the `appIDPMembership == nil` branch to confirm `app_idp: ""` and `app_directories: []` are present in the returned struct.
3. Minor: consider a one-liner comment at `appAuth["app_directories"] = directoriesData` explaining it's always set for state consistency.

---

### Summary

Solid fixes with good cleanup. The `request_parameters` type change is well-reasoned and the HCL formatting fix is genuinely important for the import tool. The main gap is a missing write-path test for the reflection assignment of the new `*string` field. Recommend addressing that before merge.

**Review by ak-amitpathak — APPROVED** [review]
